packages/calypso-e2e/src/lib/flows/login-flow.ts

```diff
@@ -20,37 +20,68 @@ export class LoginFlow {
 	/**
 	 * Creates an instance of the log in flow.
 	 *
-	 * @param {Page} page Instance of a Playwright page on which test steps are executing.
+	 * @param {Page} page Object representing the base page.
 	 * @param {string} [accountType] Type of account to be used for the log in process.
 	 */
 	constructor( page: Page, accountType = 'defaultUser' ) {
 		this.page = page;
 
-		// Ignoring the last portion of the destructure (siteURL) because
-		// it is not useful to us yet.
 		const [ username, password ] = getAccountCredential( accountType );
 		this.username = username;
 		this.password = password;
 	}
 
 	/**
-	 * Executes the basic log in flow as the specified user.
+	 * Executes the common steps of logging in as a particular user.
+	 * Typically, this method should not be called directly by the test writer as
+	 * other login methods perform additional checks and waits.
 	 *
+	 * @param {Page} page Page on which the login interactions should occur. This is not necessarily
+	 * the same page as the base page where the test is being executed.
 	 * @returns {Promise<void>} No return value.
 	 */
-	async baseflow(): Promise< void > {
-		console.log( 'Logging in as ' + this.username );
-		const loginPage = new LoginPage( this.page );
+	async baseflow( page?: Page ): Promise< void > {
+		console.log( '\tLogging in as ' + this.username );
+
+		// Unless a page object has explictly been passed into this method, assume the login process
+		// can execute on the same page object that was passed into the constructor of this flow.
+		if ( ! page ) {
+			page = this.page;
+		}
+
+		const loginPage = await LoginPage.Expect( page );
 		await loginPage.login( { username: this.username, password: this.password } );
 	}
 
+	/* Log in abstraction methods */
+
 	/**
-	 * Log in as the user without performing any additional steps.
+	 * Log in as the specified user from the WPCOM Log-In endpoint.
+	 * This is the most basic action of logging in.
 	 *
 	 * @returns {Promise<void>} No return value.
 	 */
 	async login(): Promise< void > {
+		await this.page.goto( getCalypsoURL( 'log-in' ) );
 		await this.baseflow();
-		await MyHomePage.Expect( this.page );
+	}
+
+	/**
+	 * Log in as the specified user from a popup, typically triggered while logged out.
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async loginFromPopup(): Promise< void > {
+		// Popup emits the event 'popup'. Capturing the event obtains the Page object
+		// for the popup page, where the login form is located.
+		const popupPage = await this.page.waitForEvent( 'popup' );
```

Comment:
Because we're expecting the `popup` event here, does it mean that `loginFromPopup` will always have to be called within the `Promise.all` chain? I.e.

```js
await Promise.all( [
  loginFromPopup(),
  this.page.click( '.login-button' ),
] );
```

Reply:
Interestingly, yes and no!

The [documentation](https://playwright.dev/docs/multi-pages/#handling-popups) covers two cases: when the action that triggers the popup is known, and unknown.

If it's known then one would use the `Promise.all` chain.

My implementation here is the second case, where I treat the action that causes the popup to be 'unknown'. In that case, Playwright can still listen to the `popup` event emitted and that would grab the page handle referencing the popup page. Pretty neat.

I chose this approach because of a couple of reasons:
- generic: I wanted the `LoginFlow.loginFromPopup()` to be able to stand on its own, for it to handle all login steps on its own.
- loose coupling: I did not want the pages or tests to be tasked with the process of triggering the popup, wait for the page, then pass the page to LoginFlow for executing the action.

As always, with that said, I am open to changing the approach here! So far this approach seems to work just fine, but if I run into issues later on I'd be open to delegating the popup handling to the page/specs.

Reply:
>My implementation here is the second case, where I treat the action that causes the popup to be 'unknown'. In that case, Playwright can still listen to the popup event emitted and that would grab the page handle referencing the popup page. Pretty neat.

Thanks for the explanation! This makes that approach totally valid, although still needs specific handling either way. Maybe we could just link to the `popups` doc to make folks aware about a proper usage of this method (known vs unknown trigger)?

Reply:
Added link to the documentation.